### PR TITLE
Compile master

### PR DESCRIPTION
**Pull request by Contentmaudlin**
Some refactoring and code fixing, but this PR is mostly for getting master to compile again

### CODE REVIEW

File: src/backend/semant.ml [hunk 1]
```diff
@@ -133,7 +111,7 @@ and check_expr tr_env expr =
    | Id(var) -> 
       if VarMap.mem var tr_env.scope
 			then let x = List.hd (VarMap.find var tr_env.scope) in x.var_type
-			else raise (Failure ("Variable " ^ var ^ "has not been declared"))
+			else raise (Failure ("Variable " ^ var ^ " has not been declared"))
```
Comment (lucas-schuermann) [inline src/backend/semant.ml:132]: <3 for being pedantic

**Review by lucas-schuermann — APPROVED** [review]

**Comment by lucas-schuermann** [comment]
LGTM--some areas seem to still be in progress. Going to verify that compilation works before merging

**Review by lucas-schuermann — DISMISSED** [review]
duh--didnt read entry point enough.

this is produced by a basic main with a bool inside

`boolboolboolbool; ModuleID = 'shux'`

File: build/Makefile [hunk 2]
```diff
@@ -18,10 +18,13 @@ symlink:
 	$(SYMLINK) ../src/frontend/exceptions.ml exceptions.ml
 	$(SYMLINK) ../src/frontend/ast.mli ast.ml
 	$(SYMLINK) ../src/backend/sast.mli sast.ml
+	$(SYMLINK) ../src/backend/cast.mli cast.ml
```
Comment (j-hui) [inline build/Makefile:21]: ideally we should not be doing this. `.mli` files shouldn't have functions

File: src/backend/semant.ml [hunk 3]
```diff
@@ -15,11 +15,9 @@ module VarSet = Set.Make(struct
       let compare x y = Pervasives.compare x y
 		end)
 
-(* represent typed variables with same name *)
-(*TODO: How are we handling nested name declarations of different types?
-ideally we should disallow this *) 
 type var = {
-	id : string; 
+	id : string;
```
Comment (j-hui) [inline src/backend/semant.ml:19]: indentation sindentation

File: src/backend/semant.ml [hunk 4]
```diff
@@ -193,9 +189,25 @@ and check_expr tr_env expr =
...
+                let v = List.hd (VarMap.find l tr_env.scope) in
+                if v.mut = Mutable then v.var_type 
+                else raise (Failure "Cannot assign to immutable type")
+        in
+   (match e1 with 
+       | Id l -> ignore (match_typ e1 e2); get_mutability l
+       | Assign(l1,l2) -> match_typ e1 e2
+       | Access(x,y)-> match_typ e1 e2 
+       | _ -> raise (Failure "Assign can only be done against an id or struct field")
+   )                (*TODO: check for mutability *)
```
Comment (j-hui) [inline src/backend/semant.ml:209]: actually complete

**Review by j-hui — APPROVED** [review]
bilbo baggins says yes i do

**Review by j-hui — APPROVED** [review]
changes made

**Pull request merged**